### PR TITLE
rustc: Lower link args to `@`-files on Windows more

### LINKED ISSUE #46999: emscripten linking with many dependencies broken on Windows
Hello.

I have been trying to compile a rather big project for wasm32-unknown-emscripten using cargo on Windows.
Unfortunately, the linking process fails when trying to run emcc.bat because the command line invoked is too long for Windows' limitations (about 8k characters).
```error: linking with `emcc.bat` failed: exit code: 1```
```note: The command line is too long.```

After manually counting, the command indeed is too big with in my case about 11k characters, as it includes the absolute path of every single .o and .rlib files. My project, while including piston as a dependency, does not have an incredibly large amount of them, which makes me believe this could be an important issue in the long run (this does not apply to Linux however as the maximum character limit usually is much higher).
After a little bit of digging, emscripten seems to provide a workaround for this problem by accepting parameters from files. However it has to be handled by the user.

Thanks in advance for your help.

### PR DESCRIPTION
**Pull request by alexcrichton**
When spawning a linker rustc has historically been known to blow OS limits for
the command line being too large, notably on Windows. This is especially true of
incremental compilation where there can be dozens of object files per
compilation. The compiler currently has logic for detecting a failure to spawn
and instead passing arguments via a file instead, but this failure detection
only triggers if a process actually fails to spawn.

Unfortunately on Windows we've got something else to worry about which is
`cmd.exe`. The compiler may be running a linker through `cmd.exe` where
`cmd.exe` has a limit of 8192 on the command line vs 32k on `CreateProcess`.
Moreso rustc actually succeeds in spawning `cmd.exe` today, it's just that after
it's running `cmd.exe` fails to spawn its child, which rustc doesn't currently
detect.

Consequently this commit updates the logic for the spawning the linker on
Windows to instead have a heuristic to see if we need to pass arguments via a
file. This heuristic is an overly pessimistic and "inaccurate" calculation which
just calls `len` on a bunch of `OsString` instances (where `len` is not
precisely the length in u16 elements). This number, when exceeding the 6k
threshold, will force rustc to always pass arguments through a file.

This strategy should avoid us trying to parse the output on Windows of the
linker to see if it successfully spawned yet failed to actually sub-spawn the
linker. We may just be passing arguments through files a little more commonly
now...

The motivation for this commit was a recent bug in Gecko [1] when beta testing,
notably when incremental compilation was enabled it blew out the limit on
`cmd.exe`. This commit will also fix #46999 as well though as emscripten uses a
bat script as well (and we're blowing the limit there).

[1]: https://bugzilla.mozilla.org/show_bug.cgi?id=1430886

Closes #46999

### CODE REVIEW

**Comment by alexcrichton** [comment]
I'm also tagging this as beta-nominated since incremental is in beta now

File: src/librustc_trans/back/command.rs [hunk 1]
```diff
@@ -105,6 +105,43 @@ impl Command {
...
+        //
+        // Turns out though that `cmd.exe` has even smaller limits, 8192
+        // characters [2]. Linkers can often be batch scripts (for example
+        // Emscripten, Gecko's current build system) which means that we're
+        // running through batch scripts. These linkers often just forward
+        // arguments elsewhere (and maybe tack on more), so if we blow 8192
+        // bytes we'll typically cause them to blow as well.
+        //
+        // Basically as a result just perform an inflated estimate of what our
+        // command line will look like and test if it's > 8192. If all else
```
Comment (Mark-Simulacrum) [inline src/librustc_trans/back/command.rs]: This seems to be false? The code below compares with `6*1024`, which is 6,144. Though it's already fairly clear what we're trying to say.

**Comment by michaelwoerister** [comment]
Thank you, Alex!

r=me with the inconsistency that @Mark-Simulacrum mentioned ironed out.

File: src/librustc_trans/back/command.rs (hunk 1 above)
Comment (alexcrichton) [reply src/librustc_trans/back/command.rs]: Oh that's actually intentional in the sense that we are sort of checking against 8192, I just randomly chose 6k instead to assume some percentage of slop when calculating the length of the command line (I've updated the comment though to hopefully reflect this!)

**Comment by alexcrichton** [comment]
@bors: p=1

(this'll get backported to beta once merged)

**Comment by alexcrichton** [comment]
@bors: retry

* https://github.com/rust-lang/rust/issues/46903

**Comment by alexcrichton** [comment]
```
[01:56:46] test [run-make] run-make\long-linker-command-lines-cmd-exe has been running for over 60 seconds
```

on second thought, not spurious

**Comment by shepmaster** [comment]
```
[01:49:03] thread '[run-make] run-make\long-linker-command-lines-cmd-exe' panicked at 'explicit panic', tools\compiletest\src\runtest.rs:2884:9
```

Seems like it would be related to the contents of this PR

**Comment by alexcrichton** [comment]
I've rebased but this isn't ready for merging, I'm investigating that last failure.

**Pull request merged**